test/integration/replicaset/replicaset_test.go

```diff
@@ -268,250 +510,438 @@ func TestAdoption(t *testing.T) {
+	}
+
+	// test to ensure apps/v1beta2 selector is immutable
+	rsV1beta2, err := clientSet.AppsV1beta2().ReplicaSets(ns.Name).Get(replicaset.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to get apps/v1beta2 replicaset %s: %v", replicaset.Name, err)
+	}
+	newSelectorLabels = map[string]string{"changed_name_apps_v1beta2": "changed_test_apps_v1beta2"}
+	rsV1beta2.Spec.Selector.MatchLabels = newSelectorLabels
+	rsV1beta2.Spec.Template.Labels = newSelectorLabels
+	_, err = clientSet.AppsV1beta2().ReplicaSets(ns.Name).Update(rsV1beta2)
+	if err == nil {
+		t.Fatalf("failed to provide validation error when changing immutable selector when updating apps/v1beta2 replicaset %s", rsV1beta2.Name)
+	}
+	expectedErrType := "Invalid value"
+	expectedErrDetail := "field is immutable"
+	if !strings.Contains(err.Error(), expectedErrType) || !strings.Contains(err.Error(), expectedErrDetail) {
+		t.Errorf("error message does not match, expected type: %s, expected detail: %s, got: %s", expectedErrType, expectedErrDetail, err.Error())
 	}
 }
 
-func waitRSStable(t *testing.T, clientSet clientset.Interface, rs *v1beta1.ReplicaSet, ns string) {
-	rsClient := clientSet.Extensions().ReplicaSets(ns)
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		updatedRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+func TestSpecReplicasChange(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-spec-replicas-change", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	defer close(runControllerAndInformers(rm, informers))
+
+	rs := newRS("rs", ns.Name, 2)
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
+	waitRSStable(t, c, rs)
+
+	// Update .Spec.Replicas and verify .Status.Replicas is changed accordingly
+	scaleRS(t, c, rs, 3)
+	scaleRS(t, c, rs, 0)
+	scaleRS(t, c, rs, 2)
+
+	// Add a template annotation change to test RS's status does update
+	// without .Spec.Replicas change
+	rsClient := c.Extensions().ReplicaSets(ns.Name)
+	updateRS(t, rsClient, rs, func(rs *v1beta1.ReplicaSet) {
+		rs.Spec.Template.Annotations = map[string]string{"test": "annotation"}
+	})
+	newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to update rs %s: %v", rs.Name, err)
+	}
+	if newRS.Generation <= rs.Generation {
+		t.Fatalf("Failed to verify .Generation has incremented for rs %s: %v", rs.Name, err)
+	}
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newRS, err = rsClient.Get(rs.Name, metav1.GetOptions{})
 		if err != nil {
 			return false, err
 		}
-		if updatedRS.Status.Replicas != *rs.Spec.Replicas {
-			return false, nil
-		} else {
-			return true, nil
-		}
+		return newRS.Status.ObservedGeneration > rs.Generation, nil
```

Comment:
You need to save the value of `newRS.Generation` from above, and compare against that. Also, you should expect `Status.ObservedGeneration >= savedGeneration` because it might be equal.

Reply:
Agreed. With correct `savedGeneration` value, there is a possibility for `ObservedGeneration` to match it.